### PR TITLE
[receiver/prometheus] syncTargetAllocator now detects regex changes in relabel config

### PR DESCRIPTION
**Pull request by rashmichandrashekar**
**Description:** <Describe what has changed.>
Fixing a bug - With the Targetallocator component in prometheus receiver, when prometheus scrape config is updated for metric relabel config with just regex change, the prometheus metrics receiver doesn't update the hash and hence doesn't pick up the metrics relabel config with the new regex.
This is because the **regex** struct has unexported fields. This fix similar to the fix made in [opentelemetry-operator](https://github.com/open-telemetry/opentelemetry-operator/pull/2171) fixes this issue by taking into account the unexported fields as well.

**Link to tracking Issue:** - #29313 

**Testing:** Tested to make sure that just the regex changes in metric relabeling gets picked up with TargetAllocator enabled.

Reopening this PR since the old one(https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30258) got closed due to inactivity.

### CODE REVIEW

**Comment by rashmichandrashekar** [comment]
@Aneurysm9 @dashpole - Please take a look and let me know if you have any questions/concerns. Thanks!

**Comment by rashmichandrashekar** [comment]
> @Aneurysm9 @dashpole - Please take a look and let me know if you have any questions/concerns. Thanks!

@Aneurysm9 @dashpole - Gentle ping. Checking back. Pls take a look and let me know if you have any feedback. Thanks!

**Comment by rashmichandrashekar** [comment]
> > @Aneurysm9 @dashpole - Please take a look and let me know if you have any questions/concerns. Thanks!
> 
> @Aneurysm9 @dashpole - Gentle ping. Checking back. Pls take a look and let me know if you have any feedback. Thanks!

@Aneurysm9 - I have addressed the feedback from the sig meeting. Please take a look. Thanks!

**Comment by rashmichandrashekar** [comment]
> > > @Aneurysm9 @dashpole - Please take a look and let me know if you have any questions/concerns. Thanks!
> > 
> > 
> > @Aneurysm9 @dashpole - Gentle ping. Checking back. Pls take a look and let me know if you have any feedback. Thanks!
> 
> @Aneurysm9 - I have addressed the feedback from the sig meeting. Please take a look. Thanks!

@Aneurysm9 - Gentle ping, pls take a look and let me know if you have any feedback. Thanks!

**Comment by rashmichandrashekar** [comment]
> > > > @Aneurysm9 @dashpole - Please take a look and let me know if you have any questions/concerns. Thanks!
> > > 
> > > 
> > > @Aneurysm9 @dashpole - Gentle ping. Checking back. Pls take a look and let me know if you have any feedback. Thanks!
> > 
> > 
> > @Aneurysm9 - I have addressed the feedback from the sig meeting. Please take a look. Thanks!
> 
> @Aneurysm9 - Gentle ping, pls take a look and let me know if you have any feedback. Thanks!

@dashpole/ @Aneurysm9  - Could you please take a look at the PR when you get a chance? Thanks!

**Comment by rashmichandrashekar** [comment]
> > > > > @Aneurysm9 @dashpole - Please take a look and let me know if you have any questions/concerns. Thanks!
> > > > 
> > > > 
> > > > @Aneurysm9 @dashpole - Gentle ping. Checking back. Pls take a look and let me know if you have any feedback. Thanks!
> > > 
> > > 
> > > @Aneurysm9 - I have addressed the feedback from the sig meeting. Please take a look. Thanks!
> > 
> > 
> > @Aneurysm9 - Gentle ping, pls take a look and let me know if you have any feedback. Thanks!
> 
> @dashpole / @Aneurysm9 - Could you please take a look at the PR when you get a chance since its been pending for quite sometime? Thanks!

**Comment by swiatekm-sumo** [comment]
@rashmichandrashekar could you please rebase your branch, so your PR only has changes to the prometheus receiver? Right now, it has a bunch of other random changes.

**Comment by rashmichandrashekar** [comment]
> @rashmichandrashekar could you please rebase your branch, so your PR only has changes to the prometheus receiver? Right now, it has a bunch of other random changes.

Thanks @swiatekm-sumo. It is done. The changes now are all relevant to the go mod and go sum for the hashstructure module removal.

**Comment by rashmichandrashekar** [comment]
@djaglowski - Following up. Could you please take a look at this PR? I see that you are assigned to this. Thanks!

File: receiver/prometheusreceiver/internal/structhash.go
Comment (swiatekm-sumo) [inline receiver/prometheusreceiver/internal/structhash.go:1]: Do we really need this whole hashing implementation just for this? Was there a problem with the implementation from the operator itseif? That one was much simpler.

Comment (rashmichandrashekar) [reply receiver/prometheusreceiver/internal/structhash.go:1]: @swiatekm-sumo - Please see this comment - https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30258#pullrequestreview-1817059446.
There was a feedback about sorting, so that approach wouldn't work well unless we made changes. So I used the OSS library which implemented this already.

Comment (swiatekm-sumo) [reply receiver/prometheusreceiver/internal/structhash.go:1]: Why can't we just sort the keys in that loop? goyaml is otherwise deterministic in its outputs.

**Comment by rashmichandrashekar** [comment]
@dashpole - Gentle ping. Could you please review when you get a chance? Thanks!

File: receiver/prometheusreceiver/internal/structhash.go
Comment (rashmichandrashekar) [reply receiver/prometheusreceiver/internal/structhash.go:1]: @swiatekm-sumo - Thanks for reviewing. Yes, that could be done as well. Since there was an implementation already, I picked this library since it was already doing it and we discussed this in the SIG meeting a couple weeks ago. If you and @dashpole/@Aneurysm9 agree I can change the implementation to do that instead.

Comment (swiatekm-sumo) [reply receiver/prometheusreceiver/internal/structhash.go:1]: If the maintainers of this component are ok with maintaining this hashing code specifically for this purpose, I guess I have no objections to that.

Comment (rashmichandrashekar) [reply receiver/prometheusreceiver/internal/structhash.go:1]: Thanks @swiatekm-sumo. @dashpole @Aneurysm9 - Please let me know your thoughts. Thanks!

**Comment by rashmichandrashekar** [comment]
> @dashpole - Gentle ping. Could you please review when you get a chance? Thanks!

@dashpole / @Aneurysm9 - Ping. Please let me know if there's any other feedback. We have been waiting on this to be fixed to take the new collector release. Could you also please provide an ETA if possible as to when this might be looked into?

**Comment by rashmichandrashekar** [comment]
> > @dashpole - Gentle ping. Could you please review when you get a chance? Thanks!
> 
> @dashpole / @Aneurysm9 - Ping. Please let me know if there's any other feedback. We have been waiting on this to be fixed to take the new collector release. Could you also please provide an ETA if possible as to when this might be looked into?

@djaglowski - I see you assigned to this PR. Could you pls take a look and let me know your feedback. Thanks!

**Comment by dashpole** [comment]
I'm not sure I'm willing to accept this. I don't want to own a hashing library.  We are working on splitting out the target allocator into its own module in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33146 so it can have separate owners.  If the future owners of the module want to own a hashing library, thats up to them.

**Comment by swiatekm-sumo** [comment]
@rashmichandrashekar it looks like we might be moving the target allocator related code to its own package, as per #33146. For that reason, your PR might need to wait until that's decided. Apologies for the wait, and thank you for your patience.

**Comment by rashmichandrashekar** [comment]
> @rashmichandrashekar it looks like we might be moving the target allocator related code to its own package, as per #33146. For that reason, your PR might need to wait until that's decided. Apologies for the wait, and thank you for your patience.

Thanks @swiatekm-sumo for letting me know. Please let me know when the move is done and owners are decided and I can update the PR with any feedback from the owners of the code.

**Comment by rashmichandrashekar** [comment]
> I'm not sure I'm willing to accept this. I don't want to own a hashing library. We are working on splitting out the target allocator into its own module in #33146 so it can have separate owners. If the future owners of the module want to own a hashing library, thats up to them.

Thanks @dashpole. I will follow up with @swiatekm-sumo / future owners to get feedback and update the PR.

**Comment by rashmichandrashekar** [comment]
> > @rashmichandrashekar it looks like we might be moving the target allocator related code to its own package, as per #33146. For that reason, your PR might need to wait until that's decided. Apologies for the wait, and thank you for your patience.
> 
> Thanks @swiatekm-sumo for letting me know. Please let me know when the move is done and owners are decided and I can update the PR with any feedback from the owners of the code.

@swiatekm-sumo / @dashpole - Are the new owners for the targetallocator component figured out? I can follow up with the right owners while the code is being split to address any comments.

**Comment by swiatekm-sumo** [comment]
@rashmichandrashekar yeah, in my opinion we should do something very similar to what the target allocator does and just serialize the configs to yaml, and hash the output. They should be sorted, to avoid map ordering inconsistencies.

File: receiver/prometheusreceiver/metrics_receiver.go
Comment (swiatekm-sumo) [inline receiver/prometheusreceiver/metrics_receiver.go:152]: Could we add some simple unit tests for this function? There was concern about map iteration order inconsistency, so we should at least check if this function ran on two copies of the same config gives the same result.

File: receiver/prometheusreceiver/metrics_receiver_target_allocator_test.go
Comment (swiatekm-sumo) [inline receiver/prometheusreceiver/metrics_receiver_target_allocator_test.go:649]: If we do need to wait here, could we use `assert.Eventually` instead?

**Review by swiatekm — COMMENTED** [review]
This approach looks good to me. I left some comments about testing.

File: receiver/prometheusreceiver/metrics_receiver_target_allocator_test.go
Comment (rashmichandrashekar) [reply receiver/prometheusreceiver/metrics_receiver_target_allocator_test.go]: @swiatekm - This is not needed. Removed it.

File: receiver/prometheusreceiver/metrics_receiver.go
Comment (rashmichandrashekar) [reply receiver/prometheusreceiver/metrics_receiver.go]: Thanks @swiatekm. Added one. Please take a look.

**Review by swiatekm — APPROVED** [review]
LGTM

**Comment by rashmichandrashekar** [comment]
> LGTM

Thanks @swiatekm for the quick review and approval. @dashpole - Could you pls help get this merged?

**Comment by rashmichandrashekar** [comment]
> > LGTM
> 
> Thanks @swiatekm for the quick review and approval. @dashpole - Could you pls help get this merged?

@dashpole - Following up, could you pls help get this merged as this is waiting for several months?

**Comment by dashpole** [comment]
Ack.  I'm OK with the general approach now.  I'll try to review soon

**Review by dashpole — APPROVED** [review]

**Pull request merged**
